src/tokens2.jl

```diff
@@ -0,0 +1,163 @@
+import Base.isless
+import Base.isequal
+import Base.colon
+import Base.endof
+
+
+SortedContainer = Union(SortedDict, SortedMultiDict, SortedSet)
+SortedContainer1 = Union(SortedDict, SortedMultiDict)
+typealias Token @compat Tuple{SortedContainer, IntSemiToken}
+typealias Token1 @compat Tuple{SortedContainer1, IntSemiToken}
+typealias SetToken @compat Tuple{SortedSet, IntSemiToken}
+
+## Function startofs returns the semitoken that points
+## to the first sorted order of the tree.  It returns
+## the past-end token if the tree is empty.
+
+@inline startof(m::SortedContainer) = IntSemiToken(beginloc(m.bt))
+
+## Function endofs returns the semitoken that points
+## to the last item in the sorted order,
+## or the before-start marker if the tree is empty.
+
+@inline endof(m::SortedContainer) = IntSemiToken(endloc(m.bt))
+
+## Function pastendsemitoken returns the token past the end of the data.
+
+@inline pastendsemitoken(::SortedContainer) = IntSemiToken(2)
+
+## Function beforestarttoken returns the token before the start of the data.
+
+@inline beforestartsemitoken(::SortedContainer) = IntSemiToken(1)
+
+## delete! deletes an item given a token.
+
+@inline function delete!(ii::Token)
+    has_data(ii)
+    delete!(ii[1].bt, ii[2].address)
+end
+
+## Function advances takes a token and returns the
+## next token in the sorted order. 
+
+@inline function advance(ii::Token)
+    not_pastend(ii)
+    IntSemiToken(nextloc0(ii[1].bt, ii[2].address))
+end
+
+
+## Function regresss takes a token and returns the
+## previous token in the sorted order. 
+
+@inline function regress(ii::Token)
+    not_beforestart(ii)
+    IntSemiToken(prevloc0(ii[1].bt, ii[2].address))
+end
+
+
+## status of a token is 0 if the token is invalid, 1 if it points to 
+## ordinary data, 2 if it points to the before-start location and 3 if
+## it points to the past-end location.
+
+
+@inline status(ii::Token) =
+       !(ii[2].address in ii[1].bt.useddatacells)? 0 :
+                        (ii[2].address == 1? 2 : (ii[2].address == 2? 3 : 1))
+
+
+@inline compare(m::SortedContainer, s::IntSemiToken, t::IntSemiToken) = 
+    compareInd(m.bt, s.address, t.address)
+
+
+@inline function deref(ii::Token1)
+    has_data(ii)
+    return ii[1].bt.data[ii[2].address].k, ii[1].bt.data[ii[2].address].d
+end
+
+@inline function deref(ii::SetToken)
+    has_data(ii)
+    return ii[1].bt.data[ii[2].address].k
+end
+
+@inline function deref_key(ii::Token1)
+    has_data(ii)
+    return ii[1].bt.data[ii[2].address].k
+end
+
+@inline function deref_value(ii::Token1)
+    has_data(ii)
+    return ii[1].bt.data[ii[2].address].d
+end
+
+
+## Functions setindex! and getindex for semitokens:
+
+@inline function getindex(m::SortedContainer1, 
+                          ::Type{IntSemiToken}, 
+                          i::IntSemiToken)    
```

Comment:
Since you already specified the type for `i`, the `::Type{IntSemiToken}` is unnecessary, as I suggested above, unless the user is using `SemiToken`s as keys (which would be really confusing).

I can see that `m[i]` and `m[t]` might be confusing if `i` were a key and `t` were a (semi-)token--is that why you chose to implement it this way?

At any rate, this construct is not very "Julian" (for `getindex`--this type of signature is used pretty commonly for other types of functions), so when using both for indexing, I think it would be better to suggest that the user use descriptive variable names--`m[i]` vs `m[token]` isn't too bad.

Am I missing something?